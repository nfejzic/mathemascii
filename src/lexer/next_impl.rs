macro_rules! next_impl {
    ($self: ident, no_prefix: $($infallible_fn:ident),*; prefix: $($fallible_fn:ident),* or $fallback_fn:ident) => {{
        let mut curr = $self.curr;
        let mut token = None;

        $(
        if let Some((num, cursor)) = $self.$infallible_fn() {
            // number is never prefix of anything
            $self.curr = cursor;
            return Some(num);
        }
        )*

        $(
        if let Some((found_token, cursor)) = $self.$fallible_fn(curr - $self.curr + 1) {
            token = Some(found_token);
            curr = cursor;

            // token can't contain a whitespace, so this token has the maximum length already
            if $self.src.get(cursor).map(|s| s.is_whitespace()).unwrap_or(false) {
                $self.curr = curr;
                return token;
            }
        }
        )*

        match token {
            Some(token) => {
                $self.curr = curr;
                Some(token)
            }
            None => {
                let (token, curr) = $self.$fallback_fn(curr - $self.curr + 1)?;
                $self.curr = curr;
                return Some(token);
            }
        }
    }};
}

pub(super) use next_impl;
