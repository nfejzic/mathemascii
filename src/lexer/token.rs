use super::{
    keywords::{
        accents::Accent, arrows::Arrow, font_commands::FontCommand, functions::Function,
        greeks::Greek, groupings::Grouping, logicals::Logical, operators::Operator, others::Other,
        relations::Relation,
    },
    Span,
};

/// Token identified in the ascii math source.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) struct Token<'src> {
    kind: TokenKind,
    content: &'src str,
    span: Span,
}

impl Token<'_> {
    pub fn with_span(content: &str, kind: TokenKind, span: Span) -> Token<'_> {
        Token {
            content,
            kind,
            span,
        }
    }

    #[allow(dead_code)] // is used in testing
    pub fn span(&self) -> Span {
        self.span
    }

    #[allow(dead_code)] // is used in testing
    pub fn kind(&self) -> TokenKind {
        self.kind
    }

    #[allow(dead_code)] // is used in testing
    pub fn as_str(&self) -> &str {
        self.content
    }
}

/// Kind of token identified in ascii math input.
#[derive(Default, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) enum TokenKind {
    /// Number literal.
    Number,

    /// Greek letter, e.g. pi, alpha etc.
    Greek(Greek),

    /// Various arrows, for example |->
    Arrow(Arrow),

    /// Standard functions, eg. sin, cos etc.
    Function(Function),

    /// Standard operators, e.g. +, -, *, |>< etc.
    Operator(Operator),

    /// Symbols that are not letters, and aren't explicitely defined in AsciiMath grammar fall back
    /// to unknown operators.
    UnknownOperator,

    /// Relations in maths, e.g. =, !=, <, <= etc.
    Relation(Relation),

    /// Logic symbols, e.g. and, or, for all etc.
    Logical(Logical),

    /// Grouping brackets in ascii math, e.g. (, {, [, (: etc.
    Grouping(Grouping),

    /// Other (Miscellaneous) symbols of ascii math.
    Other(Other),

    /// Ascii math keywords for accents, e.g. hat, bar, ubrace etc.
    Accent(Accent),

    /// Ascii math font commands such as bb, cc etc.
    FontCommand(FontCommand),

    /// Other Ascii symbol that was not recognized as a token.
    Variable,

    #[default]
    Unimplemented,
}

impl TokenKind {
    pub fn is_grouping_open(&self) -> bool {
        let TokenKind::Grouping(grp) = self else {
            return false;
        };

        !matches!(
            grp,
            Grouping::CloseParen
                | Grouping::CloseBracket
                | Grouping::CloseBrace
                | Grouping::RightAngled
                | Grouping::CloseIgnored
        )
    }
}
