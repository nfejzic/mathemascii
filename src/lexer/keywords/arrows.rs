use crate::lexer::token::TokenKind;

use super::macros::generate_impl;

generate_impl!(
    Arrow,
    Arrows,
    "uarr" | "uparrow" => Up,
    "darr" | "downarrow" => Down,
    "->" | "to" | "rarr" | "rightarrow" => Right,
    ">->" | "rightarrowtail" => RightTail,
    "->>" | "twoheadrightarrow" => TwoHeadRight,
    ">->>" | "twoheadrightarrowtail" => TwoHeadRightTail,
    "|->" | "mapsto" => MapsTo,
    "larr" | "leftarrow" => Left,
    "harr" | "leftrightarrow" => LeftRight,
    "rArr" | "Rightarrow" => BigRight,
    "lArr" | "Leftarrow" => BigLeft,
    "hArr" | "Leftrightarrow" => BigLeftRight,
    prefixes:
        RightTail => ">->>",
        TwoHeadRight => "twoheadrightarrowtail",
        Right => "->>"
);

impl From<Arrow> for TokenKind {
    fn from(value: Arrow) -> Self {
        TokenKind::Arrow(value)
    }
}
