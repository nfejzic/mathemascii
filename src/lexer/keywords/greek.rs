use crate::lexer::token::TokenKind;

use super::macros::generate_impl;

generate_impl!(
    Greek,
    Greeks,
    "alpha" => Alpha,
    "beta" => Beta,
    "gamma" => Gamma,
    "Gamma" => BigGamma,
    "delta" => Delta,
    "epsilon" => Epsilon,
    "varepsilon" => Varepsilon,
    "zeta" => Zeta,
    "eta" => Eta,
    "theta" => Theta,
    "vartheta" => Vartheta,
    "iota" => Iota,
    "kappa" => Kappa,
    "lambda" => Lambda,
    "mu" => Mu,
    "nu" => Nu,
    "xi" => Xi,
    "Xi" => BigXi,
    "pi" => Pi,
    "Pi" => BigPi,
    "rho" => Rho,
    "sigma" => Sigma,
    "Sigma" => BigSigma,
    "tau" => Tau,
    "upsilon" => Upsilon,
    "phi" => Phi,
    "Phi" => BigPhi,
    "varphi" => Varphi,
    "chi" => Chi,
    "psi" => Psi,
    "Psi" => BigPsi,
    "omega" => Omega,
    "Omega" => BigOmega
);

impl From<Greek> for TokenKind {
    fn from(value: Greek) -> Self {
        TokenKind::Greek(value)
    }
}
