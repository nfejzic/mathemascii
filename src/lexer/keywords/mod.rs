//! Constants containing all the keywords defined in the asciimath grammar.

use crate::scanner::Symbol;

use super::token::TokenKind;

mod macros;

pub mod arrows;
pub mod functions;
pub mod greek;

pub(crate) trait KeywordKind: Into<TokenKind> + Copy {
    fn prefix_of(&self) -> Option<usize>;
}

pub(crate) trait Keyword {
    const MAX_LEN: usize;
    const MIN_LEN: usize;

    type Kind: KeywordKind;

    fn get(key: &str) -> Option<Self::Kind>;
    fn starts_with(symbol: Symbol<'_>) -> bool;
}
