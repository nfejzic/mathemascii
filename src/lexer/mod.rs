//! Tokens, lexer and other helper types and functions for tokenization of asciimath input.

mod next_impl;
mod token;

use crate::scanner::{Scan, Symbol};

use token::{Token, TokenKind};

use self::keywords::{
    accents::Accents, arrows::Arrows, functions::Functions, greek::Greeks, groupings::Groupings,
    logicals::Logicals, operators::Operators, others::Others, relations::Relations, Keyword,
    KeywordKind,
};

mod keywords;

#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) struct Span {
    start: usize,
    end: usize,
}

/// Iterator that finds and returns tokens in AsciiMath input.
///
/// In cases where a token is prefix of other token, the longer token is given precedence. For
/// example: 'g' is function g, and 'gamma' is greek letter. In order to correctly identify the
/// greek letter, the longer token must have precedence.
#[derive(Default, Debug, Clone, PartialEq, Eq)]
pub(crate) struct TokenIterator<'src> {
    src: Vec<Symbol<'src>>,
    curr: usize,
    span: Span,
}

pub(crate) trait Tokenize {
    fn tokenize(&self) -> TokenIterator;
}

impl<T> Tokenize for T
where
    T: Scan,
{
    fn tokenize(&self) -> TokenIterator {
        TokenIterator {
            src: self.scan(),
            ..Default::default()
        }
    }
}

impl<'src> TokenIterator<'src> {
    fn skip_whitespace(&mut self) {
        loop {
            match self.src.get(self.curr) {
                Some(sym) if sym.is_whitespace() => self.curr += 1,
                _ => break,
            }
        }
    }

    /// Identifies a number in AsciiMath input, like 42, 42.24, .3 etc.
    fn lex_number(&self) -> Option<(Token<'src>, usize)> {
        let mut dot_seen = false;

        let start = self.curr;
        let mut curr = self.curr;

        while let Some(sym) = self.src.get(curr) {
            if !sym.is_digit() && !sym.is_dot() {
                break;
            } else if sym.is_dot() {
                if dot_seen {
                    break;
                }

                dot_seen = true;
            }

            curr += 1;
        }

        if start == curr {
            return None;
        }

        let content = Symbol::as_str(self.src.get(start..curr)?)?;

        if matches!(content, ".") {
            // single dot is not a valid number
            return None;
        }

        let kind = TokenKind::Number;

        let span = Span { start, end: curr };
        Some((Token::with_span(content, kind, span), curr))
    }

    /// Lexes a keyword with a given minimum length (or default for the given keyword if that is
    /// longer)
    fn lex_keyword<K>(&self, min_len: usize) -> Option<(Token<'src>, usize)>
    where
        K: Keyword,
    {
        if let Some(sym) = self.src.get(self.curr) {
            if !<K as Keyword>::starts_with(*sym) {
                // none of the corresponding keywords start with the given symbol, so skip parsing
                return None;
            }
        }

        let min_len = K::MIN_LEN.max(min_len);

        let start = self.curr;
        let mut curr = (self.curr + min_len).saturating_sub(1);
        let mut found_at = start;
        let mut keyword = None;

        while let Some(slice) = self.src.get(start..=curr) {
            if slice.last().map(|s| s.is_whitespace()).unwrap_or(false) {
                // token can't contain a whitespace
                break;
            }

            curr += 1;

            let len = slice.len();

            if len > K::MAX_LEN {
                break;
            }

            let slice_str = Symbol::as_str(slice)?;

            if let Some(kind) = <K as Keyword>::get(slice_str) {
                // longer keywords have precedence, otherwise they would not be possible to lex...
                let span = Span { start, end: curr };
                keyword = Some(Token::with_span(slice_str, kind.into(), span));
                found_at = curr;

                match kind.prefix_of() {
                    Some(word_len) => {
                        if len > word_len {
                            break;
                        }
                    }
                    None => break,
                }
            }
        }

        // keyword lexed up to the `found_at` index. Since we check if longer keywords can be
        // lexed, current index is set beyond this point. To make sure we don't overshoot the
        // index, reset to the position where the keyword was actually lexed.
        keyword.map(|k| (k, found_at))
    }

    /// Identifies a greek letter in AsciiMath input, e.g. alpha, beta, pi, Psi etc.
    fn lex_greek(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Greeks>(min_len)
    }

    /// Identifies an arrow in AsciiMath input, e.g. ->, MapsTo, |-> etc.
    fn lex_arrow(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Arrows>(min_len)
    }

    /// Identifies a function in AsciiMath input, e.g. log, sin, cosh etc.
    fn lex_function(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Functions>(min_len)
    }

    fn lex_operator(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Operators>(min_len)
    }

    fn lex_relation(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Relations>(min_len)
    }

    fn lex_logical(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Logicals>(min_len)
    }

    fn lex_grouping(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Groupings>(min_len)
    }

    fn lex_other(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Others>(min_len)
    }

    fn lex_accent(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Accents>(min_len)
    }
}

impl<'src> Iterator for TokenIterator<'src> {
    type Item = Token<'src>;

    fn next(&mut self) -> Option<Self::Item> {
        self.skip_whitespace();

        match self.src.get(self.curr) {
            Some(_) => {
                next_impl::next_impl!(
                self,
                no_prefix:
                    lex_number;
                prefix:
                    lex_greek,
                    lex_arrow,
                    lex_function,
                    lex_operator,
                    lex_relation,
                    lex_logical,
                    lex_grouping,
                    lex_other,
                    lex_accent
                )
            }
            None => None,
        }
    }
}

#[cfg(test)]
mod tests;
