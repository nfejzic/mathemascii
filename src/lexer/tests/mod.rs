use crate::lexer::{
    keywords::greek::Greek,
    token::{Token, TokenKind},
    Tokenize,
};

#[test]
fn skip_whitespace() {
    let src = "   alpha   24.42";
    let tokens: Vec<_> = src.tokenize().collect();

    assert_eq!(
        tokens,
        vec![
            Token::new("alpha", TokenKind::Greek(Greek::Alpha)),
            Token::new("24.42", TokenKind::Number),
        ]
    );
}

mod number {
    use super::super::*;

    #[test]
    fn only_number() {
        let src = "24.42";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), Some(Token::new("24.42", TokenKind::Number)));
    }

    #[test]
    fn start_with_dot() {
        let src = ".42";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), Some(Token::new(".42", TokenKind::Number)));

        let src = ".42a";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), Some(Token::new(".42", TokenKind::Number)));
    }

    #[test]
    fn start_and_stop_dot() {
        let src = ".42.3";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), Some(Token::new(".42", TokenKind::Number)));
    }

    #[test]
    fn stop_at_non_digit() {
        let src = "24.42a";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), Some(Token::new("24.42", TokenKind::Number)));
    }

    #[test]
    fn stop_at_dot() {
        let src = "24.42.";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), Some(Token::new("24.42", TokenKind::Number)));
    }
}

mod greek {
    use crate::lexer::keywords::greek::Greek;

    use super::super::*;

    #[test]
    fn greek_pi() {
        let src = "pi";
        let mut tokens = src.tokenize();
        assert_eq!(
            tokens.next(),
            Some(Token::new("pi", TokenKind::Greek(Greek::Pi)))
        );

        let src = "Pi";
        let mut tokens = src.tokenize();
        assert_eq!(
            tokens.next(),
            Some(Token::new("Pi", TokenKind::Greek(Greek::BigPi)))
        );
    }

    #[test]
    fn greek_alpha() {
        let src = "alpha";
        let mut tokens = src.tokenize();
        assert_eq!(
            tokens.next(),
            Some(Token::new("alpha", TokenKind::Greek(Greek::Alpha)))
        );
    }

    #[test]
    fn too_long() {
        let src = "abcdefghijklmnopqrstuvwxy";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), None);
    }

    #[test]
    fn multiple() {
        let src = "alphabetagammadelta";

        let tokens: Vec<_> = src.tokenize().collect();

        assert_eq!(
            tokens,
            vec![
                Token::new("alpha", TokenKind::Greek(Greek::Alpha)),
                Token::new("beta", TokenKind::Greek(Greek::Beta)),
                Token::new("gamma", TokenKind::Greek(Greek::Gamma)),
                Token::new("delta", TokenKind::Greek(Greek::Delta)),
            ]
        );
    }
}

mod arrow {
    use crate::lexer::keywords::arrows::Arrow;

    use super::super::*;

    #[test]
    fn right() {
        let src = "rarr rightarrow";
        let tokens: Vec<_> = src.tokenize().collect();
        assert_eq!(
            tokens,
            vec![
                Token::new("rarr", TokenKind::Arrow(Arrow::Right)),
                Token::new("rightarrow", TokenKind::Arrow(Arrow::Right)),
            ]
        );
    }

    #[test]
    fn big_left() {
        let src = "lArrLeftarrow";
        let tokens: Vec<_> = src.tokenize().collect();

        assert_eq!(
            tokens,
            vec![
                Token::new("lArr", TokenKind::Arrow(Arrow::BigLeft)),
                Token::new("Leftarrow", TokenKind::Arrow(Arrow::BigLeft)),
            ]
        );
    }

    #[test]
    fn too_long() {
        let src = "abcdefghijklmnopqrstuvwxy";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), None);
    }

    #[test]
    fn multiple() {
        let src = "uarr->|->twoheadrightarrowtailto";

        let tokens: Vec<_> = src.tokenize().collect();

        assert_eq!(
            tokens,
            vec![
                Token::new("uarr", TokenKind::Arrow(Arrow::Up)),
                Token::new("->", TokenKind::Arrow(Arrow::To)),
                Token::new("|->", TokenKind::Arrow(Arrow::MapsTo)),
                Token::new(
                    "twoheadrightarrowtail",
                    TokenKind::Arrow(Arrow::TwoHeadRightTail)
                ),
                Token::new("to", TokenKind::Arrow(Arrow::To)),
            ]
        );
    }
}

mod function {
    use crate::lexer::keywords::functions::Function;

    use super::super::*;

    #[test]
    fn sin_cos() {
        let src = "sincos";
        let tokens: Vec<_> = src.tokenize().collect();
        assert_eq!(
            tokens,
            vec![
                Token::new("sin", TokenKind::Function(Function::Sin)),
                Token::new("cos", TokenKind::Function(Function::Cos)),
            ]
        );
    }

    #[test]
    fn f_g() {
        let src = "fg";
        let tokens: Vec<_> = src.tokenize().collect();
        assert_eq!(
            tokens,
            vec![
                Token::new("f", TokenKind::Function(Function::F)),
                Token::new("g", TokenKind::Function(Function::G)),
            ]
        );
    }

    #[test]
    fn ln_log() {
        let src = "ln logln";
        let tokens: Vec<_> = src.tokenize().collect();
        assert_eq!(
            tokens,
            vec![
                Token::new("ln", TokenKind::Function(Function::Ln)),
                Token::new("log", TokenKind::Function(Function::Log)),
                Token::new("ln", TokenKind::Function(Function::Ln)),
            ]
        );
    }

    #[test]
    fn too_long() {
        let src = "abcdefghijklmnopqrstuvwxy";
        let mut tokens = src.tokenize();
        assert_eq!(tokens.next(), None);
    }

    #[test]
    fn multiple() {
        let src = "sincoshfg";

        let tokens: Vec<_> = src.tokenize().collect();

        assert_eq!(
            tokens,
            vec![
                Token::new("sin", TokenKind::Function(Function::Sin)),
                Token::new("cosh", TokenKind::Function(Function::CosH)),
                Token::new("f", TokenKind::Function(Function::F)),
                Token::new("g", TokenKind::Function(Function::G)),
            ]
        );
    }
}
