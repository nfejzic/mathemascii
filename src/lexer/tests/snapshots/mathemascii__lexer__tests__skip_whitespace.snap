---
source: src/lexer/tests/mod.rs
expression: "Snapshot((input, tokens))"
---
   alpha   24.42

   alpha
   ^^^^^ -> Greek(Alpha) at: 3 -> 8

           24.42
           ^^^^^ -> Number at: 11 -> 16


